airflow/models/baseoperator.py

```diff
@@ -207,6 +207,7 @@ def apply_defaults(self: "BaseOperator", *args: Any, **kwargs: Any) -> Any:
 
             result = func(self, **kwargs, default_args=default_args)
             # Store the args passed to init -- we need them to support task.map serialzation!
+            kwargs.pop('task_id', None)
```

Comment:
I don’t get what this implies.

Reply:
Delete the task_id form the kwargs (so it isn't stored in the mapped data structure) -- so `del kwargs['task_id']` but without the try/except for when it's not there? I'll double check if I can just do this with a `del`.

Reply:
I guess my question is more like why do we need to treat `task_id` specially (and why we didn’t need to previous)

Reply:
Task ID was getting stored in partial_kwargs which I hadn't noticed until I looked at the serialized representation of a MappedOperator.

Reply:
How about storing `task_id` separately instead and not in `partial_kwargs`?

Reply:
Oh right I remember the problem now. Because we end up calling apply_defaults multiple times (once for each class in the MRO) task_id gets added in to kwargs for the `super().__init__(**kwargs)` call, so this del here is to make sure it doesn't get stored in partial kwargs.

Reply:
Hmm shouldn’t this be fixed on `partial_kwargs` instead then? I’m surprised this doesn’t cause problems elsewhere. Or is `task_id` being present here wrong in the first place and should be generally fixed? (In that case we should add a comment.)

Reply:
This is dealing with it in partial_kwargs -- 

```
            # Store the args passed to init -- we need them to support task.map serialzation!
            kwargs.pop('task_id', None)
            self._BaseOperator__init_kwargs.update(kwargs)  # type: ignore
```

Reply:
This is the other option:

```diff
diff --git a/airflow/models/baseoperator.py b/airflow/models/baseoperator.py
index 752b8a6cf..71170320d 100644
--- a/airflow/models/baseoperator.py
+++ b/airflow/models/baseoperator.py
@@ -136,6 +136,7 @@ class BaseOperatorMeta(abc.ABCMeta):
         non_optional_args = {
             name for (name, param) in non_variadic_params.items() if param.default == param.empty
         }
+        non_optional_args -= {'task_id'}
 
         class autostacklevel_warn:
             def __init__(self):
@@ -158,7 +159,7 @@ class BaseOperatorMeta(abc.ABCMeta):
             func.__globals__['warnings'] = autostacklevel_warn()
 
         @functools.wraps(func)
-        def apply_defaults(self: "BaseOperator", *args: Any, **kwargs: Any) -> Any:
+        def apply_defaults(self: "BaseOperator", *args: Any, task_id: str, **kwargs: Any) -> Any:
             from airflow.models.dag import DagContext
             from airflow.utils.task_group import TaskGroupContext
 
@@ -201,15 +202,15 @@ class BaseOperatorMeta(abc.ABCMeta):
 
             hook = getattr(self, '_hook_apply_defaults', None)
             if hook:
-                args, kwargs = hook(**kwargs, default_args=default_args)
+                args, kwargs = hook(task_id=task_id, **kwargs, default_args=default_args)
+                task_id = kwargs.pop('task_id')
                 default_args = kwargs.pop('default_args', {})
 
             if not hasattr(self, '_BaseOperator__init_kwargs'):
                 self._BaseOperator__init_kwargs = {}
 
-            result = func(self, **kwargs, default_args=default_args)
+            result = func(self, **kwargs, task_id=task_id, default_args=default_args)
             # Store the args passed to init -- we need them to support task.map serialzation!
-            kwargs.pop('task_id', None)
             self._BaseOperator__init_kwargs.update(kwargs)  # type: ignore
 
             # Here we set upstream task defined by XComArgs passed to template fields of the operator

```

Which do you think is better @uranusjr?

Reply:
The “other” one looks better to me, it feels like we are intentionally treating `task_id` different from all other arguments.

Reply:
Done in https://github.com/apache/airflow/pull/20743/commits/9439378183cb4da70e33e9c6f04d2d52924709a5 -- and a bit of a justification/reason in the commit message

Reply:
Turns out this has an unfortunate side effect:

```pycon
>>> FooOperator()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: apply_defaults() missing 1 required keyword-only argument: 'task_id'
```

and error message seems to be hard-coded in the interpreter and cannot be patched. I’m trying yet another approach (see new commits)